SQL_DDL_Func_Mariam_Poghosyan_HW.sql

```diff
@@ -0,0 +1,196 @@
+            payment p
+        JOIN rental r ON p.rental_id = r.rental_id
+        JOIN inventory i ON r.inventory_id = i.inventory_id
+        JOIN film f ON i.film_id = f.film_id
+        JOIN film_category fc ON f.film_id = fc.film_id
+        JOIN category c ON fc.category_id = c.category_id
+        WHERE 
+            EXTRACT(YEAR FROM p.payment_date) = %s AND
+            EXTRACT(QUARTER FROM p.payment_date) = %s
+        GROUP BY c.name
+        HAVING SUM(p.amount) > 0
+    $f$, current_year, current_qtr);
+
+    RAISE NOTICE 'View "sales_revenue_by_category_qtr" created successfully for Year: %, Quarter: %', current_year, current_qtr;
+END;
+$$ LANGUAGE plpgsql;
+
+
+SELECT create_sales_revenue_by_category_qtr_view();
+
+
+--2
+CREATE OR REPLACE FUNCTION get_sales_revenue_by_category_qtr(qtr INT, yr INT)
+RETURNS TABLE (
+    category TEXT,
+    total_sales_revenue NUMERIC
+) AS $$
+    SELECT 
+        c.name AS category,
+        SUM(p.amount) AS total_sales_revenue
+    FROM 
+        payment p
+    JOIN rental r ON p.rental_id = r.rental_id
+    JOIN inventory i ON r.inventory_id = i.inventory_id
+    JOIN film f ON i.film_id = f.film_id
+    JOIN film_category fc ON f.film_id = fc.film_id
+    JOIN category c ON fc.category_id = c.category_id
+    WHERE 
+        EXTRACT(YEAR FROM p.payment_date) = yr AND
+        EXTRACT(QUARTER FROM p.payment_date) = qtr
+    GROUP BY c.name
+    HAVING SUM(p.amount) > 0
+$$ LANGUAGE sql STABLE;
+
+SELECT * FROM get_sales_revenue_by_category_qtr(2, 2024);
+
+--3
+CREATE OR REPLACE FUNCTION most_popular_films_by_countries(countries TEXT[])
+RETURNS TABLE (
+    country_name TEXT,
+    film_title TEXT,
+    rental_count INTEGER
+) AS $$
+BEGIN
+    RETURN QUERY
+    SELECT 
+        co.country,
+        f.title,
+        COUNT(*) AS rental_count
+    FROM 
+        rental r
+    JOIN inventory i ON r.inventory_id = i.inventory_id
+    JOIN film f ON i.film_id = f.film_id
+    JOIN customer cu ON r.customer_id = cu.customer_id
+    JOIN address a ON cu.address_id = a.address_id
+    JOIN city ci ON a.city_id = ci.city_id
+    JOIN country co ON ci.country_id = co.country_id
+    WHERE co.country = ANY(countries)
+    GROUP BY co.country, f.title
+    HAVING COUNT(*) = (
+        SELECT MAX(film_rentals.count_per_film)
+        FROM (
+            SELECT 
+                f2.title,
+                COUNT(*) AS count_per_film
+            FROM 
+                rental r2
+            JOIN inventory i2 ON r2.inventory_id = i2.inventory_id
+            JOIN film f2 ON i2.film_id = f2.film_id
+            JOIN customer cu2 ON r2.customer_id = cu2.customer_id
+            JOIN address a2 ON cu2.address_id = a2.address_id
+            JOIN city ci2 ON a2.city_id = ci2.city_id
+            JOIN country co2 ON ci2.country_id = co2.country_id
+            WHERE co2.country = co.country
+            GROUP BY f2.title
+        ) AS film_rentals
+    );
+END;
+$$ LANGUAGE plpgsql STABLE;
+
+
+
+--4

```

Comment:
Your function works incorrectly. Incorrect row_number. Moreover you need to show only last customer for each film.
![image](https://github.com/user-attachments/assets/6b7cfcb5-3381-4c9b-bd20-3117a8ef80d8)